### PR TITLE
win_package: Clean up parameters, bugfix, docs

### LINKED ISSUE #3067: win_package not accepting exit code 3010 as success
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

win_package
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

Default options
##### OS / ENVIRONMENT

Ansible on RHEL 6, deploying to a Windows 2008 R2 server
##### SUMMARY

When installing some msi, the return code is 3010 (requires restart). This is considered as a failed return code instead of a successful return code
##### STEPS TO REPRODUCE

Call the win_package module for an installer that returns exit code 3010 when installing (Visual C++ 2008 redistributable for example)

```
- name: Install Visual C++ 2008 Runtime
  win_package:
    name: 'Microsoft Visual C++ 2008 Runtime'
    path: 'C;\temp\vcredist_x86.exe'
    product_id: '{FF66E9F6-83E7-3A3E-AF14-8DE9A809A6A4}'
    arguments: '/q'
```
##### EXPECTED RESULTS

The step should be successful and changed as the package was installed. As an example, chocolatey accept several return codes as valid. It could also be implemented by having a new property to list valid return codes.
##### ACTUAL RESULTS

returned json:
{
  "_ansible_no_log": false,
  "changed": false,
  "failed": true,
  "msg": "The return code 3010 was not expected. Configuration is likely not correct",
  "name": "Microsoft Visual C++ 2008 Runtime",
  "play": "Deploy VC++",
  "event_loop": null,
  "task": "vcredist : Install Visual C++ 2008 Runtime",
  "role": "vcredist",
  "ignore_errors": null,
  "status": "failed",
  "module_name": "win_package",
  "created": "2016-09-29T13:55:09.750Z",
  "host_id": 852,
  "host_name": "targetserv",
  "id": 101457,
  "parent": 101456,
  "event": "Host Failed"
}

### PR DESCRIPTION
**Pull request by dagwieers**
##### SUMMARY
This PR includes:
- Used Get-AnsibleParam and strict typing for module parameters
- Added check-mode support
- Added `-aliases` support
- Renamed `product_id` to `productid`
- Renamed `user_name` to `username`
- Renamed `user_password` to `password`
- Clean up documentation
- Improve return information

This fixes ansible/ansible-modules-extras#3067

##### ISSUE TYPE

 - Bugfix Pull Request
 - Docs Pull Request

##### COMPONENT NAME

win_package

##### ANSIBLE VERSION

v2.4

### CODE REVIEW

**Comment by dagwieers** [comment]
@nitzmahone @jborean93 @jhawkesworth @trondhindenes Ready to be reviewed and merged.

**Comment by dagwieers** [comment]
Wake up call to to WWG :-) @nitzmahone @jborean93 @jhawkesworth @trondhindenes

File: lib/ansible/modules/windows/win_package.py [hunk 1]
```diff
@@ -25,69 +25,59 @@
...
-      - Location of the package to be installed (either on file system, network share or url)
+    - Location of the package to be installed (either on file system, network share or URL).
     required: true
   name:
     description:
-      - Name of the package, if name isn't specified the path will be used for log messages
-    required: false
-    default: null
+    - Name of the package, if name is not specified the path will be used as name.
   product_id:
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_package.py]: Isn't it the opposite, this should be `productid` and the alias is `product_id`?

File: lib/ansible/modules/windows/win_package.py [hunk 2]
```diff
@@ -125,7 +115,7 @@
   win_package:
     path: https://download.microsoft.com/download/1/6/7/167F0D79-9317-48AE-AEDB-17120579F8E2/NDP451-KB2858728-x86-x64-AllOS-ENU.exe
     productid: '{7DEBE4EB-6B40-3766-BB35-5CBBC385DA37}'
-    arguments: '/q /norestart'
-    ensure: present
-    expected_return_code: [0,3010]
+    arguments: /q /norestart
+    state: present
+    expected_return_code: [0, 3010]
 '''
```
Comment (jborean93) [inline lib/ansible/modules/windows/win_package.py]: Are you able to add the return values for this?

**Review by jborean93 — CHANGES_REQUESTED** [review]
Are you able to add check mode tests to the integration tests as well if you have added the support.

I can see the need to review this module and win_msi at some point and come up with an easier solution when installing programs.

File: lib/ansible/modules/windows/win_package.py (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_package.py]: You are right, fixed.

File: lib/ansible/modules/windows/win_package.py (hunk 2 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_package.py]: Sure, missed it.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/26064/9/tests

**Comment by trondhindenes** [comment]
I'm not sure I see the point in just "brutally" changing param names. People use these modules in production, so we should be _very_ careful / have good reasons for just ripping out params and replacing them. We have alias and depcrecation stuff available, so lets use those to make sure we give users a smooth transition.

**Comment by dagwieers** [comment]
@trondhindenes There are aliases, so there is no impact. The change in naming is to bring these in line with other modules and get more conformity between Windows modules themselves and the other non-Windows modules. This is going to make it much easier for people that don't have to remember that **win_package** uses *user_name* and *user_password* uniquely, contrary to standard practice.

There may only be a problem for people using the new parameters and then going back to an older version of this module. But that's the case for so many things, we are not forward-compatible.

PS There is a ticket for improving these things and it was discussed once during a core meeting. ansible/ansible#20160

**Comment by mattclay** [comment]
CI failure in integration tests `test/integration/targets/win_package/tasks/main.yml:51`:

```
{
    "assertion": "not win_package_install_again_result|failed", 
    "changed": false, 
    "evaluated_to": false, 
    "failed": true
}
```

**Comment by dagwieers** [comment]
@mattclay Not sure what is going on there, the previous task clearly did not fail. Still the assert fails on the fact that it not failed...

**Comment by mattclay** [comment]
@dagwieers The problem is caused by the module result:

```
ok: [windows-2012-R2_RTM] => {
    "failed": false,
    "win_package_install_again_result": {
        "arguments": null,
        "changed": false,
        "failed": false,
        "name": "C:\\Program Files\\7z922-x64.msi",
        "path": "C:\\Program Files\\7z922-x64.msi",
        "productid": "{23170F69-40C1-2702-0922-000001000000}",
        "rc": null,
        "restart_required": false
    }
}
```

Specifically, the presence of `rc: null` causes the failure.

**Comment by dagwieers** [comment]
@mattclay Hmm, the `failed` filter plugin still makes assumptions about the `rc` return value. That's not right, we wanted to get rid of that. Let's see what the impact is if we change that behaviour...

(Not sure whether win_package should return $null in this case, we added this to ensure people using that value don't have to test if `rc` was returned. Don't like it that much, but it is convenient.)

**Comment by mattclay** [comment]
CI failure in integration tests:

```
{
    "arguments": null, 
    "changed": false, 
    "failed": true, 
    "msg": "This error A parameter with the name 'WhatIf' was defined multiple times for the command.", 
    "name": "C:\\Program Files\\7z922-x64.msi", 
    "path": "C:\\Program Files\\7z922-x64.msi", 
    "productid": "{23170F69-40C1-2702-0922-000001000000}", 
    "rc": null, 
    "restart_required": false
}
```

File: lib/ansible/modules/windows/win_package.ps1 [hunk 3]
```diff
@@ -1253,77 +1258,62 @@ namespace Source
...
-    }
-    catch
-    {
+if (-not (Test-TargetResource @dscparams)) {
+    try {
+        Set-TargetResource @dscparams
+    } catch {
         $errormsg = $_
-        Fail-Json -obj $result -message $errormsg.ToString()
+        Fail-Json -obj $result -message "This error $errormsg"
```
Comment (nitzmahone) [inline lib/ansible/modules/windows/win_package.ps1]: Suspect this is debug cruft? (just noticed it in the test failure)

**Review by nitzmahone — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_package.ps1 (hunk 3 above)
Comment (dagwieers) [reply lib/ansible/modules/windows/win_package.ps1]: Yes, need to figure out where in the code it's doing -WhatIf twice, because I do not understand it.

Comment (dagwieers) [reply lib/ansible/modules/windows/win_package.ps1]: So this code has 2 issues now that I cannot explain:
- It reports a failure on rc=null (even when the task did not appear to have failed, and we explicitly set failed=false
- Somehow Set-TargetResource reports having -WhatIf provided twice (which is not possible).

**Comment by mattclay** [comment]
CI failure with the same `WhatIf` error as before.

**Comment by mattclay** [comment]
CI failure in integration tests due to:

```
{
"arguments": null, 
"changed": false, 
"failed": true, 
"msg": "A parameter with the name 'WhatIf' was defined multiple times for the command.", 
"name": "C:\\Program Files\\7z922-x64.msi", 
"path": "C:\\Program Files\\7z922-x64.msi", 
"productid": "{23170F69-40C1-2702-0922-000001000000}", 
"rc": null, 
"restart_required": false
}
```

**Comment by jborean93** [comment]
@dagwieers, are you find with me closing this PR now that the win_package rewrite has been merged?

**Comment by dagwieers** [comment]
Fine by me. Thanks !

**Pull request closed without merging**